Add 3.10.1.0 changelogs and remove the used up tiny changlog files (#8831)
haskell/cabal: release-notes/Cabal-3.10.1.0.md

Cabal and Cabal-syntax 3.10.1.0 changelog and release notes
---

Release 3.10.1.0 of cabal now sets the `--enable-documentation` option by default
when running `cabal haddock` (https://github.com/haskell/cabal/issues/7462).

> If that new default does not fit your workflow, pass an explicit `--disable-documentation`

Reviewer: (or this)

> If that new default does not fit your workflow, pass an explicit `--disable-documentation`

Author: I went for "the".

option or consider (helping to implement) the other workarounds proposed

> in https://github.com/haskell/cabal/issues/8725.

Author: I don't understand. The link you mention is on the last line above.

> in https://github.com/haskell/cabal/issues/8725.

Reviewer: sorry, I meant that all links in the .md are in the form: square-bracket ticket-number square-bracket link-between-parentheses.

This one is a “naked” link, not in the usual (for the document) form.

> in https://github.com/haskell/cabal/issues/8725.

Reviewer 2: They render okay if you view them on GitHub ([check it out](https://github.com/Mikolaj/cabal/blob/changelogs-of-3.10.1.0/release-notes/Cabal-3.10.1.0.md)). 

If you view them in plain text, you are better off with naked. 

It only leaves the case when people view the notes in some MD previewer, which I'd expect to be a very rare case.

> in https://github.com/haskell/cabal/issues/8725.

Author: @ffaf1: so, what's your final recommendation?

> in https://github.com/haskell/cabal/issues/8725.

Reviewer: Oh I did not notice. Very good then.

> 

Author: Could you elaborate?

> 

Author: I got it, thanks.

This version of cabal introduces support for JS cross-compilation (https://github.com/haskell/cabal/pull/8636).
We've seen related Windows segfaults on CI, so please kindly report if you can reproduce either
cabal or GHC segfaults when cross-compiling for JS or compiling normally but with any files
listed in the `js-sources` field of a .cabal file.

This release of cabal also fixes a lot of bugs in cabal 3.8.1.0 and not all of the fixes
are listed here.

### Significant changes

- Add support for the XDG Base Directory Specification [#680](https://github.com/haskell/cabal/issues/680) [#7386](https://github.com/haskell/cabal/pull/7386)

  Cabal/cabal-install now uses the XDG Base Directory Specification to
  store configuration, caches, and the store.  Specifically,
  `$XDG_CONFIG_HOME/cabal` stores the configuration file,
  `$XDG_CACHE_HOME/cabal` stores downloaded packages and similar, and
  `$XDG_STATE_HOME/cabal` mainly contains the store of compiled

>   packages.  Further, `cabal install will put executables in

Reviewer 2: There seem to be missing backtick:
```suggestion
  packages.  Further, `cabal install` will put executables in
```

Author: Well spotted.

  `~/.local/bin` by default.

  The `dist`/`dist-newstyle` directories are not affected.

  On Windows, these XDG paths are mapped to other (hopefully)
  appropriate locations.  See the Cabal User Guide for information.

  If the `CABAL_DIR` environment variable is set, the indicated
  directory will be used to store all Cabal-related files, as in
  previous versions.

>   Backwards compatibility: If `~/.cabal` already exists, this will be

Author: A good idea.

  interpreted as `CABAL_DIR=~/.cabal`.  This means that upgrading on an
  existing system already using cabal-install should not cause any
  change in behaviour.  An existing system can be migrated by deleting
  `~/.cabal` (possibly copying `~/.cabal/config` to
  `~/.config/cabal/config` first).

> - Add PackageInfo_ module [#8534](https://github.com/haskell/cabal/pull/8534)

Reviewer 2: ```suggestion
- Add `PackageInfo_` module [#8534](https://github.com/haskell/cabal/pull/8534)
```

Author: Done.

  - Add PackageInfo_ module to embed portable package-related informations (issue #3909)

> - *Cabal-syntax* Add language extensions DeepSubsumption and TypeData [#8493](https://github.com/haskell/cabal/pull/8493)
> 
>   - adds support for the DeepSubsumption language extension (GHC proposal #511)
>   - adds support for the TypeData language extension (GHC proposal #106)

Reviewer 2: ```suggestion
- *Cabal-syntax* Add language extensions `DeepSubsumption` and `TypeData` [#8493](https://github.com/haskell/cabal/pull/8493)

  - adds support for the `DeepSubsumption` language extension ([GHC proposal #511](https://github.com/ghc-proposals/ghc-proposals/pull/511))
  - adds support for the `TypeData` language extension ([GHC proposal #106](https://github.com/ghc-proposals/ghc-proposals/pull/106))
```

> - Add support for ghc-9.4s -fprof-late flag. [#8528](https://github.com/haskell/cabal/pull/8528)

Reviewer 2: ```suggestion
- Add support for GCH 9.4+ `-fprof-late` flag. [#8528](https://github.com/haskell/cabal/pull/8528)
```

>   - This adds the new late-toplevel profiling-detail option which enables `-fprof-late` based automatic cost centre annotations for ghcs which support it.

Reviewer 2: ```suggestion
  - This adds the new `late-toplevel` (and its alias `late`) profiling-detail option which enables `-fprof-late`-based automatic cost centres annotations for GHCs that support it ([earliest is GHC 9.4.1](https://downloads.haskell.org/ghc/9.4.1/docs/users_guide/profiling.html#ghc-flag--fprof-late)).
```

Author: Done.

### Other changes

- Remove "'-main-is' is not portable" check [#8646](https://github.com/haskell/cabal/issues/8646) [#8651](https://github.com/haskell/cabal/pull/8651)

>   'cabal check' no longer complains about '-main-is' flag in 'ghc-options'
> 
> - cabal init should not suggest Cabal < 2.0 [#8680](https://github.com/haskell/cabal/issues/8680)
> 
>   'cabal init' no longer suggests users to set cabal-version to less than 2.0
> 
> - Split up Distribution.Simple.Setup [#8130](https://github.com/haskell/cabal/pull/8130)

Reviewer 2: ```suggestion
  `cabal check` no longer complains about `-main-is` flag in `ghc-options`

- `cabal init` should not suggest Cabal < 2.0 [#8680](https://github.com/haskell/cabal/issues/8680)

  `cabal init` no longer suggests users to set cabal-version to less than 2.0

- Split up `Distribution.Simple.Setup` [#8130](https://github.com/haskell/cabal/pull/8130)
```

  The generation of the functions `minusFileName` and `splitFileName`
  are now in the same conditional block as their call,
  preventing generation of inconsistent Paths_ files
  where those functions are used but not defined.

- Remove warning on option -threaded when building libraries [#774](https://github.com/haskell/cabal/issues/774) [#8431](https://github.com/haskell/cabal/issues/8431) [#8432](https://github.com/haskell/cabal/pull/8432)

  - Accompanied by option `-flink-rts`, option `-threaded` defines the flavour of
    the ghc RTS library the built library will be linked against. Note that bare
    ghc does not warn when option `-threaded` is used for building a library
    either.
  - Note that the changes require modification of the regression check for issue
    #774 which can be regarded as a proper test for this PR.

- Add warning about expensive globs [#5311](https://github.com/haskell/cabal/issues/5311) [#8441](https://github.com/haskell/cabal/pull/8441)

  - Now cabal check will emit a warning when package uses
  recursive globs starting at root of the project

> - orders extra stanzas [#8458](https://github.com/haskell/cabal/issues/8458) [#8499](https://github.com/haskell/cabal/pull/8499)
> 
>   Ensure that extra-src-dirs, extra sources, and extra other modules all are added using ordNub rather than incidentally alphabetized.

Reviewer 2: ```suggestion
- Order `extra*` stanzas [#8458](https://github.com/haskell/cabal/issues/8458) [#8499](https://github.com/haskell/cabal/pull/8499)

  Ensure that `extra-src-dirs`, extra sources, and extra other modules all are added using ordNub rather than incidentally alphabetized.
```

> - prepend rather than append extra prog path [#6304](https://github.com/haskell/cabal/issues/6304) [#8506](https://github.com/haskell/cabal/pull/8506)

Reviewer 2: ```suggestion
- Prepend rather than append `extra-prog-path` [#6304](https://github.com/haskell/cabal/issues/6304) [#8506](https://github.com/haskell/cabal/pull/8506)
```

>   Prepends the extra-prog-path to the system path rather than appending, to allow binaries in the extra path to override defaults.
> 
> - don't send non-extant extra-lib-dirs to ghc [#6492](https://github.com/haskell/cabal/issues/6492) [#8510](https://github.com/haskell/cabal/pull/8510)

Reviewer 2: ```suggestion
  Prepends the `extra-prog-path` to the system path rather than appending, to allow binaries in the extra path to override defaults.

- don't send non-extant `extra-lib-dirs` to GHC [#6492](https://github.com/haskell/cabal/issues/6492) [#8510](https://github.com/haskell/cabal/pull/8510)
```

> - don't send non-extant extra-lib-dirs to ghc [#6492](https://github.com/haskell/cabal/issues/6492) [#8510](https://github.com/haskell/cabal/pull/8510)

Author: https://en.wiktionary.org/wiki/extant sounds correct to me?

  - Note that the changes require modification of the regression check for issue
    #774 which can be regarded as a proper test for this PR.

- Add warning about expensive globs [#5311](https://github.com/haskell/cabal/issues/5311) [#8441](https://github.com/haskell/cabal/pull/8441)

  - Now cabal check will emit a warning when package uses
  recursive globs starting at root of the project

- orders extra stanzas [#8458](https://github.com/haskell/cabal/issues/8458) [#8499](https://github.com/haskell/cabal/pull/8499)

  Ensure that extra-src-dirs, extra sources, and extra other modules all are added using ordNub rather than incidentally alphabetized.

- prepend rather than append extra prog path [#6304](https://github.com/haskell/cabal/issues/6304) [#8506](https://github.com/haskell/cabal/pull/8506)

  Prepends the extra-prog-path to the system path rather than appending, to allow binaries in the extra path to override defaults.

- don't send non-extant extra-lib-dirs to ghc [#6492](https://github.com/haskell/cabal/issues/6492) [#8510](https://github.com/haskell/cabal/pull/8510)

  If an extra-libs-dir does not exist, it does not get sent to ghc, which can error on windows.

> - Add support for ghc-9.4s -fprof-late flag. [#8528](https://github.com/haskell/cabal/pull/8528)

Reviewer 2: I consider this more important than the language extensions that made it into “important changes”.

Reviewer 2: Yep, that flag.

Yeah, I propose we patch

Author: Done.
